src/snowflake/cli/_plugins/snowpark/commands.py

```diff
@@ -164,29 +170,33 @@ def deploy(
         raise ClickException(msg)
 
     # Create stage
-    stage_name = snowpark.stage_name
-    stage_manager = StageManager()
-    stage_name = FQN.from_string(stage_name).using_context()
-    stage_manager.create(fqn=stage_name, comment="deployments managed by Snowflake CLI")
-
     snowflake_dependencies = _read_snowflake_requrements_file(
         paths.snowflake_requirements_file
     )
+    stage_names = {
+        entity.stage for entity in [*functions.values(), *procedures.values()]
+    }
+    stage_manager = StageManager()
 
-    artifact_stage_directory = get_app_stage_path(stage_name, snowpark.project_name)
-    artifact_stage_target = (
-        f"{artifact_stage_directory}/{paths.artifact_file.path.name}"
-    )
+    # TODO: Raise error if stage name is not provided
 
-    stage_manager.put(
-        local_path=paths.artifact_file.path,
-        stage_path=artifact_stage_directory,
-        overwrite=True,
-    )
+    for stage in stage_names:
+        stage = FQN.from_string(stage).using_context()
+        stage_manager.create(fqn=stage, comment="deployments managed by Snowflake CLI")
+        artifact_stage_directory = get_app_stage_path(stage, pd.defaults.project_name)
```

Comment:
You're assuming `pd.defaults.project_name` is defined, but shouldn't it be optional? What should be the stage if different values are defined on each entity?

Reply:
ATM we have only one snowpark project value. I think we should introduce different stage paths after introducing mixins, as it will change much more logic. I.e. now we create only one artifact for all snowpark entities.

Reply:
I now realize you used the `defaults` section to store global-like values for entities to reference, but it's not exactly its purpose. You can think of `defaults` as automatic mixins that are applied to all entities if they have a value of the same name and the user didn't specify it on the entity.

In this example `sp1.project_name` will be "proj1", and `sp2.project_name` will be "default_proj":
```yml
entities:
  sp1:
    type: snowpark
    project_name: proj1
  sp2:
    type: snowpark
  defaults:
    project_name: default_proj
```

Reply:
Yes, you're right- i used the default value when converting from v1 and went on with this logic.
Nevertheless - currently it doesn't change anything, as we use single artifact for all snowpark entities